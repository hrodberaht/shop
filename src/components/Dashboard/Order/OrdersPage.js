import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';

import { getOrders, getLoadedStatus } from '../../../store/orders/selectors';
import ConnectedAuthorization from '../../Auth/Authorization';

import { fetchOrders, fetchChangeOrderStatus } from '../../../store/orders/actionCreators';
import { getAuthToken, getAuthUserId } from '../../../store/authenticate/selectors';
import OrdersSearch from './OrdersSearch';
import OrdersList from './OrdersList';

export class OrdersPage extends Component {
  state = {
    status: 'all',
  };

  componentDidMount() {
    const { getOrdersFromServer, token, userId } = this.props;
    getOrdersFromServer(userId, token);
  }

  handleSelect = (e) => {
    this.setState({ status: e.target.value });
  };

  render() {
    const { orders, loaded } = this.props;
    if (!loaded) return <p>Loading...</p>;
    return (
      <React.Fragment>
        <div>
          <OrdersSearch handleSelect={this.handleSelect} />
        </div>
        <table>
          <tbody>
            <tr className="order-table-title">
              <th>Id:</th>
              <th>Person:</th>
              <th>Date:</th>
              <th>Total price:</th>
              <th>Status:</th>
              <th>Actions:</th>
              <ConnectedAuthorization render withRoleAdmin={<th>Admin:</th>} />
            </tr>
            <OrdersList filters={this.state.status} orders={orders} />
          </tbody>
        </table>
      </React.Fragment>
    );
  }
}
const mapStateToProps = state => ({
  token: getAuthToken(state),
  orders: getOrders(state),
  loaded: getLoadedStatus(state),
  userId: getAuthUserId(state),
});

export default connect(
  mapStateToProps,
  {
    handleClick: fetchChangeOrderStatus,
    getOrdersFromServer: fetchOrders,
  },
)(OrdersPage);

OrdersPage.propTypes = {
  orders: PropTypes.arrayOf(PropTypes.object).isRequired,
  loaded: PropTypes.bool.isRequired,
  getOrdersFromServer: PropTypes.func.isRequired,
  token: PropTypes.string,
  userId: PropTypes.string,
};

OrdersPage.defaultProps = {
  token: null,
  userId: null,
};
