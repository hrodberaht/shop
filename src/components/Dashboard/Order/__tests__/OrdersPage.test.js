import { OrdersPage } from '../OrdersPage';

const defaultProps = {
  orders: [],
  loaded: false,
  getOrdersFromServer: jest.fn(),
  token: '1234',
  userId: '1a2a',
};
const setup = buildSetup(OrdersPage, defaultProps);

describe('<OrdersPage />', () => {
  it('should render without crash', () => {
    const { wrapper } = setup();
    expect(wrapper).toMatchSnapshot();
  });
  it('should call getOrdersFromServer when component did mount', () => {
    const {
      wrapper,
      props: { token, userId },
    } = setup();
    expect(wrapper.instance().props.getOrdersFromServer).toHaveBeenCalledWith(userId, token);
  });
  it('state should be', () => {
    const { wrapper } = setup();
    expect(wrapper.state()).toEqual({
      status: 'all',
      id: '',
    });
  });

  it('should change state after call handleSelect', () => {
    const { wrapper } = setup();
    const e = { target: { value: 'realized' } };
    wrapper.instance().handleSelect(e);
    expect(wrapper.state('status')).toBe('realized');
  });
  it('should change state after call handleChange', () => {
    const { wrapper } = setup();
    const e = { target: { value: 'a' } };
    wrapper.instance().handleChange(e);
    expect(wrapper.state('id')).toBe('a');
  });
  it('should render if loaded true ', () => {
    const { wrapper } = setup();
    wrapper.setProps({ loaded: true });
    expect(wrapper).toMatchSnapshot();
  });
});
