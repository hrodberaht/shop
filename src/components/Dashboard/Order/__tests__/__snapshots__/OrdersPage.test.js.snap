// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<OrdersPage /> should render if loaded true  1`] = `
<Fragment>
  <div>
    <OrdersSearch
      handleChange={[Function]}
      handleSelect={[Function]}
    />
    <hr />
  </div>
  <table>
    <tbody>
      <tr
        className="order-table-title"
      >
        <th>
          Id:
        </th>
        <th>
          Person:
        </th>
        <th>
          Date:
        </th>
        <th>
          Total price:
        </th>
        <th>
          Status:
        </th>
        <th>
          Actions:
        </th>
        <Connect(Authorization)
          render={true}
          withRoleAdmin={
            <th>
              Admin:
            </th>
          }
        />
      </tr>
      <Connect(OrdersList)
        filterValues={
          Object {
            "id": "",
            "status": "all",
          }
        }
        orders={Array []}
      />
    </tbody>
  </table>
</Fragment>
`;

exports[`<OrdersPage /> should render without crash 1`] = `
<p>
  Loading...
</p>
`;
